common/djangoapps/course_modes/views.py

```diff
@@ -135,6 +136,7 @@ def get(self, request, course_id, error=None):
 
         return render_to_response("course_modes/choose.html", context)
 
+    @method_decorator(transaction.non_atomic_requests)
     @method_decorator(login_required)
```

Comment:
This decorator should not be between the two others - and it should likely be in the place where it's executed first before any transaction setup. That location would be closest to the `def post()`.